### PR TITLE
feat: Mudata support for MultiVI

### PR DESCRIPTION
**Pull request by ori-kron-wis**
Added mudata support for MULTIVI + tests

### CODE REVIEW

File: src/scvi/model/_multivi.py [hunk 1]
```diff
@@ -588,7 +589,9 @@ def get_accessibility_estimates(
             return pd.DataFrame(
                 imputed,
                 index=adata.obs_names[indices],
-                columns=adata.var_names[self.n_genes :][region_mask],
+                columns=adata["rna"].var_names[self.n_genes :][region_mask]
+                if type(adata).__name__ == "MuData"
```
Comment (canergen) [inline src/scvi/model/_multivi.py:593]: Is this the check for MuData that we use throughout the code in e.g. totalVI. We should wrap it in a helper function then.

File: src/scvi/model/_multivi.py [hunk 2]
```diff
@@ -925,130 +928,6 @@ def differential_expression(
 
         return result
 
-    @torch.no_grad()
```
Comment (canergen) [inline src/scvi/model/_multivi.py:928]: Why is this removed?

File: src/scvi/model/_multivi.py [hunk 3]
```diff
@@ -1117,3 +996,123 @@ def _check_adata_modality_weights(self, adata):
         """
         if (adata is not None) and (self.module.modality_weights == "cell"):
             raise RuntimeError("Held out data not permitted when using per cell weights")
+
+    @classmethod
+    @setup_anndata_dsp.dedent
+    def setup_mudata(
```
Comment (canergen) [inline src/scvi/model/_multivi.py:1002]: This should be default. Deprecation warning for anndata should be added and tutorial executed using MuData.

File: tests/model/test_multivi.py [hunk 4]
```diff
@@ -76,3 +85,416 @@ def test_multivi_single_batch():
...
+    # mdata.mod
+    MULTIVI.setup_mudata(
+        mdata,
+        rna_layer="counts",  # mean we use: mdata.mod["rna_subset"].layers["counts"]
+        protein_layer=None,  # mean we use: mdata.mod["protein"].X
+        batch_key="batch",  # the batch is here: mdata.mod["rna_subset"].obs["batch"]
+        modalities={
+            "rna_layer": "rna_subset",
+            "protein_layer": "protein",
+            "batch_key": "rna_subset",
```
Comment (canergen) [inline tests/model/test_multivi.py:120]: I'm confused about the batch_key. Does is need to be set up with modalities.

Comment (ori-kron-wis) [reply tests/model/test_multivi.py:120]: the batch_key under modalities actually tells which modality stores batch_key in its obs data.
so for example:

```
    MULTIVI.setup_mudata(
        mdata,
        rna_layer="counts", 
        protein_layer=None, 
        batch_key="batch",  
        modalities={
            "rna_layer": "rna_subset",
            "protein_layer": "protein",
            "batch_key": "rna_subset",
        },
    )
```

the batch is here: mdata.mod["rna_subset"].obs["batch"]

File: src/scvi/model/_multivi.py (hunk 3 above)
Comment (ori-kron-wis) [reply src/scvi/model/_multivi.py:1002]: OK. will add a deprecation warning for both this and toalvi.
Tutorial wise, maybe we will do it like in totalvi adding the setup_mudata part on top of what already existed.

File: src/scvi/model/_multivi.py (hunk 1 above)
Comment (ori-kron-wis) [reply src/scvi/model/_multivi.py:593]: fixed

File: src/scvi/model/_multivi.py [hunk 5]
```diff
@@ -959,6 +1086,12 @@ def setup_anndata(
             `adata.obsm[protein_expression_obsm_key]` if it is a DataFrame, else will assign
             sequential names to proteins.
         """
+        warnings.warn(
+            "MULTIVI is suppose to work with MuData. the use of anndata is "
```
Comment (canergen) [inline src/scvi/model/_multivi.py:1090]: correct: supposed and removed

File: src/scvi/model/_totalvi.py [hunk 6]
```diff
@@ -1215,6 +1215,12 @@ def setup_anndata(
         -------
         %(returns)s
         """
+        warnings.warn(
+            "TOTALVI is suppose to work with MuData. the use of anndata is "
+            "deprecated and will be remove in scvi-tools 1.4. Please use setup_mudata",
```
Comment (canergen) [inline src/scvi/model/_totalvi.py:1220]: I think for TOTALVI, we should deprecate it, but without announcing that it will be removed in 1.4. For MultiVI, it's a bad solution, while for TOTALVI, it's actually quite okay'ish to use obsm.

**Review by canergen — APPROVED** [review]

File: src/scvi/model/_multivi.py (hunk 5 above)
Comment (ori-kron-wis) [reply src/scvi/model/_multivi.py:1090]: fixed

File: CHANGELOG.md [hunk 7]
```diff
@@ -23,6 +23,9 @@ to [Semantic Versioning]. Full commit history is available in the
 
 #### Added
 
+- Experimental MuData support for {class}`~scvi.model.MULTIVI` via the method
```
Comment (canergen) [inline CHANGELOG.md:26]: Remove Experimental. If we deprecate setup_anndata it is stable and the recommended way.

File: tests/model/test_multivi.py [hunk 8]
```diff
@@ -76,3 +85,416 @@ def test_multivi_single_batch():
...
+    #     batch_key="batch",
+    #     modalities={"rna_layer": "rna", "protein_layer": "protein", "batch_key": "rna",
+    #                 "atac_layer": "atac"},
+    # )
+    # n_obs = mdata.n_obs
+    # n_genes = np.min([adata.n_vars, protein_adata.n_vars])
+    # n_regions = protein_adata.X.shape[1]
+    # n_latent = 10
+
+    mdata = synthetic_iid(return_mudata=True)
```
Comment (canergen) [inline tests/model/test_multivi.py:181]: Let's skip the muon testing requirement and use this also for the RNA+ATAC test?

File: tests/model/test_multivi.py [hunk 9]
```diff
@@ -76,3 +85,416 @@ def test_multivi_single_batch():
...
+        mdata,
+        batch_key="batch",
+        modalities={"rna_layer": "rna", "batch_key": "rna", "protein_layer": "protein"},
+    )
+    model = MULTIVI(mdata, n_genes=50, n_regions=50)
+    adata2 = synthetic_iid()
+    protein_adata2 = synthetic_iid(n_genes=50)
+    mdata2 = MuData({"rna": adata2, "protein": protein_adata2})
+    mdata2.obs["_indices"] = np.arange(mdata2.n_obs)
+    model.get_elbo(mdata2)
```
Comment (canergen) [inline tests/model/test_multivi.py:257]: I don't get this test. We should test all functions with mdata2

File: tests/model/test_multivi.py [hunk 10]
```diff
@@ -76,3 +85,416 @@ def test_multivi_single_batch():
...
+        batch_key="batch",
+        modalities={"rna_layer": "rna", "batch_key": "rna", "protein_layer": "protein"},
+    )
+    model = MULTIVI(mdata, n_genes=50, n_regions=50)
+    adata2 = synthetic_iid()
+    protein_adata2 = synthetic_iid(n_genes=50)
+    mdata2 = MuData({"rna": adata2, "protein": protein_adata2})
+    adata2.obs.batch = adata2.obs.batch.cat.rename_categories(["batch_1", "batch_0"])
+    mdata2.obs["_indices"] = np.arange(mdata2.n_obs)
+    model.get_elbo(mdata2)
```
Comment (canergen) [inline tests/model/test_multivi.py:300]: Same here all tests with mdata2

**Review by canergen — APPROVED** [review]
Please check whether we need muon.

File: CHANGELOG.md [hunk 11]
```diff
@@ -4,7 +4,7 @@ Starting from version 0.20.1, this format is based on [Keep a Changelog], and th
 to [Semantic Versioning]. Full commit history is available in the
 [commit logs](https://github.com/scverse/scvi-tools/commits/).
 
-## Version 1.2
+## Version 1.3
```
Comment (canergen) [inline CHANGELOG.md:8]: Remove this from the PR. It shouldn't be part of 1.2.X, whereas the rest of the commit should.

File: tests/model/test_multivi.py (hunk 10 above)
Comment (canergen) [reply tests/model/test_multivi.py:275]: This needs to be fixed. Call other functions with mdata2
